src/opencl/mod.rs

```diff
@@ -88,25 +89,75 @@ impl<T> Buffer<T> {
     }
 }
 
+#[derive(Copy, Clone, PartialEq, Eq, Hash)]
+pub struct DeviceUuid([u8; utils::CL_UUID_SIZE_KHR]);
+
+impl TryInto<DeviceUuid> for &str {
+    type Error = String;
+
+    fn try_into(self) -> Result<DeviceUuid, Self::Error> {
+        let res = self
+            .split('-')
+            .map(|s| hex::decode(s).map_err(|e| e.to_string()))
+            .collect::<Result<Vec<_>, Self::Error>>()?;
+
+        let res = res.into_iter().flatten().collect::<Vec<u8>>();
+
+        if res.len() != utils::CL_UUID_SIZE_KHR {
+            Err("Invalid string format".to_string())
+        } else {
+            let mut raw = [0u8; utils::CL_UUID_SIZE_KHR];
+            raw.copy_from_slice(res.as_slice());
+            Ok(DeviceUuid(raw))
+        }
+    }
+}
+
+impl fmt::Display for DeviceUuid {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use hex::encode;
+
+        // formats the uuid the same way as clinfo does, as an example:
+        // the output should looks like 46abccd6-022e-b783-572d-833f7104d05f
+        write!(
+            f,
+            "{}-{}-{}-{}-{}",
+            encode(&self.0[..4]),
+            encode(&self.0[4..6]),
+            encode(&self.0[6..8]),
+            encode(&self.0[8..10]),
+            encode(&self.0[10..])
+        )
+    }
+}
+
+impl fmt::Debug for DeviceUuid {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.to_string())
+    }
+}
+
 #[derive(Debug, Clone)]
 pub struct Device {
     brand: Brand,
     name: String,
     memory: u64,
-    bus_id: Option<BusId>,
     platform: ocl::Platform,
+    pci_id: Option<PciId>,
+    uuid: Option<DeviceUuid>,
     pub device: ocl::Device,
 }
 
 impl Hash for Device {
     fn hash<H: Hasher>(&self, state: &mut H) {
-        self.bus_id.hash(state);
+        self.uuid.hash(state);
```

Comment:
The UUID might not always be set, I think you need to hash the `pci_id` as well.